### DISCUSSION TITLE
Writing py5 tutorials and how-to documentation to support new creative coders

### DISCUSSION THREAD

**Discussion by marcovicci** [py5coding/py5generator#101] [Show and tell]
Hello, I'm Zelle! If you recognize that discussion title, it's the truncated form of "Writing py5 tutorials and how-to documentation to support creative coders new to Python and/or new to the Processing ecosystem", a project supported by the Processing Foundation alongside their Google Summer of Code 2022 offerings: https://medium.com/processing-foundation/announcing-google-summer-of-code-2022-projects-and-a-few-more-77043ab4d0b4 

Learning how Jupyter Notebooks functions has been strange, but I return with progress and delight (it feels really good once you actually get into it). I've made the following placeholder pages in my fork of py5book's "tutorials" folder, representing eighteen different tutorials that will hopefully get just about anybody up and running in py5. 

- Installing and using py5 (which can discuss how to use the live code functions on the documentation site, development environments like Thonny and Jupyter Notebooks with the right addons and kernels, and the difference between modes) 
- Drawing 2D primitives (coordinates, color, fills, strokes)
- Variables and data types
- Drawing complex shapes (curves, vertices, contour functions)
- Working with text (strings and the functions and methods to manipulate them, visual display of text and fonts)
- Conditional statements (if, else if, else; and, or, not)
- Iteration with loops (for, while, break, continue)
- Randomness (using random functions with iteration to create tiled patterns)
_The following pages tentatively use the py5 kernel instead of py5bot, since this is where we introduce sketches with setup() and draw(). I believe there are other syntax differences that will need to be addressed, but I'll feel it out as we go..._ 
- Animation and motion (use of non-static sketches, frames, etc)
- Mouse and keyboard interaction
- Transformations and matrixes
- Working with lists 
- Reading external data (CSV files that can generate charts)
- Dictionaries (and in addition to the standard dictionary stuff, loading in external json data)
- Custom functions (args, return, etc)
- Trigonometry for animation 
- Object-Oriented Programming
- py5vector movement (which utilizes OOP to do some really fun stuff)

Because I've been writing week-by-week, and in larger chunks than these pages represent (running through several pages worth of material each Monday), I have actual content to move into them up until Randomness and will soon have more. Not all these files will be populated in a linear order, since the structure of the class is a little different - notably, mouse and keyboard interaction will be filled in quite a bit later than the surrounding stuff. So it goes.

A good initial example of the content of these tutorials is the Conditional statements page, which has a first pass complete. (As each tutorial is moved into the correct page, I'll also be creating corresponding folders for them inside of the "images" folder, since that makes sense - and what even is py5 without visual examples?) 

https://github.com/marcovicci/py5book/blob/main/tutorials/conditional_statements.md

None of this is any good unless there are eyes on it, of course, so it's all fair game for discussion and suggestions. I'll be continuing to move content into the right places - as of writing this, the beginner coders at Massey have learned about 2D primitives, variables/data types, drawing more complex shapes, and working with text, and they've taken to it amazingly well, which makes me very optimistic!

**Comment by tabreturn** [comment]
Thanks, Zelle. I think you should also amend the https://github.com/marcovicci/py5book/blob/main/_toc.yml file, so we can see how the contents will look ordered/arranged on the site.

It'll be a long list of items to add to the contents, and I recall discussions about accordion/expanding sections to accommodate things better. Also, it'll help us consider how we best arrange the content sections.

**Reply by marcovicci** [reply]
Done, and absolutely enormous! I definitely think an accordion style menu could be our saving grace. I wouldn't want to make what we're adding look more or less significant than the existing tutorials; perhaps everything here (new and old) can end up categorized and tucked away in those sensible categories.

**Comment by marcovicci** [comment]
Added first passes at _Iteration with loops_ and _Randomness_, as well as updating https://github.com/marcovicci/py5book/blob/main/_toc.yml with this utterly huge new list of stuff. 

https://github.com/marcovicci/py5book/blob/main/tutorials/iteration_with_loops.md

https://github.com/marcovicci/py5book/blob/main/tutorials/randomness.md

**Reply by hx2A** [reply]
> Added first passes at *Iteration with loops* and *Randomness*,

I just read these and they look great! I very much like how you provide history and context to introduce the programming concepts. Your writing style is easy to read and I can see how this will be a hit with beginners.

> this utterly huge new list of stuff.

I will look into what we can do about this to keep it organized. Other projects with larger Jupyter book websites would probably provide good ideas for how we can approach this. Currently there is something accordion-like happening with each class in the Reference documentation. Maybe we can do something like that here?

**Comment by hx2A** [comment]
I like your organization, this is a good outline of all of the topics.

> (As each tutorial is moved into the correct page, I'll also be creating corresponding folders for them inside of the "images" folder, since that makes sense - and what even is py5 without visual examples?)

Let me know if you have issues with broken images, or images that work in the Jupyter notebook but aren't correct after Jupyter book gets built. The [Makefile](https://github.com/marcovicci/py5book/blob/main/Makefile) calls a [post_build.py](https://github.com/marcovicci/py5book/blob/main/scripts/post_build.py) script to make some small adjustments and corrections. Currently it isn't adjusting any image links but I recall needing to do so in the past (I can't quite remember why though).

**Reply by marcovicci** [reply]
Thanks! There might be heaps - I haven't tried building the book at all yet while I muck around figuring it all out. I'll let you know when I find some problems...

**Comment by marcovicci** [comment]
Added first passes at Animation and Motion, and Transformation and Matrices (which is now spelled correctly)...

https://github.com/marcovicci/py5book/blob/main/tutorials/animation_and_motion.md

https://github.com/marcovicci/py5book/blob/main/tutorials/transformations_and_matrices.md

These ones are weird and lengthy, and especially when dealing with matrices, I'm a bit less confident in what I'm describing (life rarely asks me to do complex maths). The students will test it all soon enough!

Animation and Motion also includes an introduction to setup() and draw(), since all other tutorials so far have been using static mode. I suspect I may have to revisit that introduction in other tutorials that require it, in case people do a lot of jumping around.

**Reply by tabreturn** [reply]
I'm thinking maybe avoid the math of matrices, @marcovicci? Just stick to the scale()/translate()/etc. funtions? Check the book content I sent you. Recall that the _Transformations_ section will follow _Animation & Motion_ in the contents for these tutorials (so something like an analogue clock example, etc. could work)?

**Reply by marcovicci** [reply]
> I'm thinking maybe avoid the math of matrices, @marcovicci? Just stick to the scale()/translate()/etc. funtions? Check the book content I sent you. Recall that the _Transformations_ section will follow _Animation & Motion_ in the contents for these tutorials (so something like an analogue clock example, etc. could work)?

Ah yeah, that'll pare it down considerably (and mean I can maybe stop muddling my way through MathJax, cool as it is) - I'll take a crack at it!

**Reply by hx2A** [reply]
@marcovicci I had never heard of "Beta movement" or "phi phenomenon". Thanks for teaching me something new!

From the animation & motion page:

> coding in py5 without using these is referred to as static mode or imported mode, depending on your environment

Unless if I misunderstand what you are writing here, I think this should only be static mode. Static mode is writing py5 code without a `setup()` or `draw()` function. Imported mode is writing code and functions without prefixing everything with `py5.`. Thonny's py5 plugin will let you write code in static mode or imported mode. The py5bot Jupyter kernel lets you write in static mode. The py5 Jupyter kernel lets you write in imported mode.

I noticed your code uses calls to the builtin `print()` function. Should we be using py5's `println()` instead? We should think about this. Beginners should certainly learn about the builtin `print()` function but it can result in output showing up all over the place when used in a Jupyter notebook. It doesn't matter when using Thonny.

http://py5.ixora.io/reference/sketch_println.html

Maybe start with `print()` and mention `println()` in a later tutorial when the difference is relevant? I'd like to hear what others think about this.

BTW, after the next release, you'll be able to specify colors with strings like 'red' or 'brown', etc. I hope to do a release in the next week or so.

**Reply by hx2A** [reply]
> Ah yeah, that'll pare it down considerably (and mean I can maybe stop muddling my way through MathJax, cool as it is) - I'll take a crack at it!

Rather than delete it, can you save it for later? What you wrote is useful, but perhaps not for a beginner tutorial. Maybe you can include some of it when you write about the Py5Vector class?

**Reply by tabreturn** [reply]
Maybe we include a `println()` _note_ where `print()` gets introduced? (right in the beginning).
```
```{note}
You'll find that these tutorials make use of `print()` and [`println()`](http://py5.ixora.io/reference/sketch_println.html) interchangeably. 
The difference ...

**Reply by marcovicci** [reply]
@hx2A 

> Unless if I misunderstand what you are writing here, I think this should only be static mode.

You're totally right - thanks for catching that! And I agree, I never chuck out work that may be useful (for now I'll let it stay where it is, but make it clear this is a reference for the mathematical basis of matrices, and later we can see if it fits better amongst vectors). 

I totally missed println() being an option; time to do more reading :)

**Reply by marcovicci** [reply]
@hx2A 

I suspect it's a silly question, but I'm just not familiar enough with jupyter - is there a way to have some blocks of code in a single tutorial run off py5bot, and some off the regular py5 kernel? Otherwise I can restructure the transformation and matrix-related stuff to use setup and draw too - but in their current form they don't need it, and the clock task they're building towards very much does, haha!

**Reply by hx2A** [reply]
> I totally missed println() being an option; time to do more reading :)

One of my important tasks is to point out the features in py5 that would be "new" and unfamiliar to folks who already know Processing.

> is there a way to have some blocks of code in a single tutorial run off py5bot, and some off the regular py5 kernel? 

Not a silly question at all. Each notebook can only have one kernel, so you must choose between the py5 kernel and the py5bot kernel. However, there are special IPython magics you can use:

http://py5.ixora.io/reference/py5magics.html

The [``%%py5bot``](http://py5.ixora.io/reference/py5magics_py5bot.html) magic is only available for the py5 kernel. If you are using the py5 kernel and you put ``%%py5bot`` at the top of a cell, it will execute that cell with py5bot. You could use this magic to combine py5bot and py5 code in a notebook that uses the py5 kernel.

However, we should ask if using magics is a good idea for beginner tutorials. The magic commands are kind of like meta-commands and are not actual Python. Is this confusing for beginners? I recall a conversation with Allison Parrish about this and she said she didn't like using magics in her Python classes for this reason.

As a side note, be aware that the P2D and P3D renderers do not work for py5bot on OSX. This is a shortcoming that is probably fixable. If you need it to be fixed for the documentation you are writing, please let me know.

**Reply by tabreturn** [reply]
_"restructure the transformation and matrix-related stuff to use setup and draw too"_ 

Maybe this is the more elegant approach here, especially if the matrix-math content moves elsewhere. After all, this section will follow the one that introduces animated sketches. It's also useful to show early on how transformations don't 'accumulate' over `draw()` frames.

**Reply by marcovicci** [reply]
First pass at Trigonometry for Animation added too. This one scares me and I'll revisit it later as well - Tristan's got some wonderful stuff with custom functions for stuff like Lissajous curves in his book that I'd like to put in there! (For now, I've kept it simple - because in the material I'm teaching, the trig stuff comes bundled with transformations and matrices, and we haven't talked custom functions yet.) 

https://github.com/marcovicci/py5book/blob/main/tutorials/trigonometry_for_animation.md

The magic is... well, magical! Thanks for that! But yes, the section on transformations and matrices can easily incorporate setup and draw, and it might be a better approach.

**Reply by tabreturn** [reply]
Looking good. A couple of things --

* It seems the unit circle image is missing? 
* At the risk of sounding like a total pedant, the midsize en dash (–) is narrower than the em dash (—) but wider than a hyphen (-). You're using hyphens throughout. For example: "triangular relationship - just draw" should be "triangular relationship — just draw". But many people think the em dash (—) is awkwardly long, so an en dash (–) is probably the way to go.
* More pedantry -- should we stick to the [PEP8 guidelines](http://pep8online.com/checkresult) (so that others contributing to documentation all adhere to the same conventions)? We might ignore the rule about two empty lines between function blocks, but I think the rest applies well here.

Also: what do you think about using shorthand hex values? So `#112233` is `#123` and so on.

**Reply by hx2A** [reply]
> The X Window System for the operating system Linux includes a feature called xeyes, with a pair of eyes that follow your mouse pointer. This is especially useful when you have multiple large displays, and might lose your cursor somewhere!

I've always considered xeyes to be this silly thing, while also losing my cursor on a 4K display. It never occurred to me that I could use xeyes to help me! I am going to start doing this. Thanks!

> You're using hyphens throughout.

I've always been confused about those longer dashes and I never knew that "en dash" and "em dash" were two separate things. Who comes up with this stuff??? One thing that really bugs me about the longer dashes is when they get mixed up in code I will copy-paste. Sometimes editors (MS Word) will be aggressive about converting hyphens to those other dashes and make code un-runnable.

How does one type an en dash in Jupyter notebook? That is not a key on my keyboard. Is there a markdown way of creating them?

> More pedantry -- should we stick to the [PEP8 guidelines](http://pep8online.com/checkresult) (so that others contributing to documentation all adhere to the same conventions)?

I think the code we present should stick to consistent guidelines because it makes the code more readable and I think some learning takes place as you get accustomed to seeing code presented in a particular way. With time, this experience makes it easier to visually spot errors and become a better coder. I've actually been able to pinpoint errors in other people's code without reading it and only seeing formatting anomalies. Formatting is actually important, but this isn't always obvious to beginners.

However, it's important to let people know we will be friendly about how these are applied, and not turn PEP8 into a bludgeon to hit people over the head with.

I'm sure you are familiar with the [spaces or tabs](https://flinhong.github.io/images/201607/tabs-spaces.jpg) cartoon where the person who mixes tabs and spaces is getting hit by the other two. After seeing the cartoon many times I realized there is something about it that makes me feel sad: the third person doesn't even know why they are being hit. The cartoon isn't funny at all. I've been that third person. Although I don't like mixing tabs and spaces (the python interpreter will reject it), and I much prefer spaces, this isn't how anyone should be treated to enforce guidelines. Coding standards should be presented in a way that is enabling and helpful when reading code, and using them part of the learning process.

**Reply by hx2A** [reply]
> Also: what do you think about using shorthand hex values? So `#112233` is `#123` and so on.

I like it! Also, after the next release, you'll be able to use any valid matplotlib color [but only if matplotlib is installed. It is not a required dependency]. So `"red"` and `"brown"` will both work.

I plan on getting the release out the door this weekend or early next week.

**Reply by hx2A** [reply]
I like how github will add a little color swatch next to hex colors: `#882222`.

**Reply by villares** [reply]
> First pass at Trigonometry for Animation added too. This one scares me and I'll revisit it later as well  

See if you like any of my trigonometry examples... 

https://github.com/villares/material-aulas/blob/main/Processing-Python-py5/seno_cosseno_atan2.md
(I haven't revised the Py.Processing to py5 conversion yet, code formatting is awful right now, as it came straight from the conversion helper)

**Reply by tabreturn** [reply]
> How does one type an en dash in Jupyter notebook?

It seems Jupyter won't render `--` as an en-dash (the width of an **n** character). As an aside, LaTeX will convert `---` to an em-dash (the width of an **m**). 

I'd probably just write all the en-dashes as as `--` for Jupyter, then use a good ol' find-and-replace to switch those out for em-dash characters.

Also, I don't mind a double-hyphen (--), but hyphen (-) just looks wrong to me.

**Reply by hx2A** [reply]
> I'd probably just write all the en-dashes as as `--` for Jupyter, then use a good ol' find-and-replace to switch those out for em-dash characters.

The py5book build process has a [post_build.py](https://github.com/py5coding/py5book/blob/main/scripts/post_build.py) script for making adjustments like this. I'll open an issue to convert all -- and --- to en dashes and em dashes.

**Reply by hx2A** [reply]
https://github.com/py5coding/py5book/issues/7

**Comment by villares** [comment]
Wow, wonderful stuff @marcovicci! Thank you all folks!

Should the binder thing work on the fork? Like this https://mybinder.org/v2/gh/marcovicci/py5book/HEAD?urlpath=lab ?

The colors part took me by surprise because I'm so used to `fill(r, g, b)` instead of the hex notation, maybe because I started on Processing Java, now I have to think if I should switch my stuff to hex!

**Reply by hx2A** [reply]
> Should the binder thing work on the fork? Like this https://mybinder.org/v2/gh/marcovicci/py5book/HEAD?urlpath=lab ?

I would expect it to work but if you are the first person to attempt this, it will be slow as it builds the image. Are you seeing an error?

**Comment by villares** [comment]
It looked like it never finished loading properly, but I'm not sure.

**Reply by hx2A** [reply]
Yes, sometimes it fails for odd reasons.

**Reply by villares** [reply]
Working now, yay!
https://hub-binder.mybinder.ovh/user/marcovicci-py5book-18jrna01/lab/tree/tutorials/color_and_shapes.ipynb

**Comment by marcovicci** [comment]
Long time no see! (Still writing, of course -- newest finished update is Mouse and keyboard interaction -- but I had a bit of a break where not much happened.) I'm back in the saddle and will be adapting the tutorial for custom functions soon. Yay!

@hx2A Tristan and I were wondering if I should do a pull request for one or two roughly finished tutorial pages? This could let us see how things are actually looking and whether something in my formatting is going to present problems overall, and give a pull request page to include in the Processing foundation's GSOC22 wrap-up post. Maybe the page on iteration/loops or randomness (as beginner introductions which use Static mode)? Or the Animation/Motion page? Let me know what you think I need to do to push them a bit farther along!

**Reply by hx2A** [reply]
Yes! I was going to ping to you see how you are doing. I'm eager to see how this is coming together. I figured you were super busy with teaching and other responsibilities. I've also been a bit swamped with family support stuff and an art project I'm working on. Should we have a video meeting soon?

I owe you an idea for how to fit the documentation you are writing into the overall documentation website. Here's a suggestion for what to put in the _toc.yml file that I think will be a good starting point for discussion.

```yml
- caption: Tutorials
  chapters:
  - file: tutorials/intro_to_py5_and_python
    sections:
    - glob: tutorials/intro_to_py5_and_python_*

  - file: tutorials/misc
    sections:
    - glob: tutorials/misc_*
```

or maybe instead:

```yml
- caption: Tutorials
  chapters:
  - file: tutorials/intro_to_py5_and_python
    sections:
    - file: tutorials/intro_to_py5_and_python_variables_and_datatypes
    - file: tutorials/intro_to_py5_and_python_randomness
    - file: tutorials/intro_to_py5_and_python_mouse_and_keyboard_interaction
    - file: tutorials/intro_to_py5_and_python_working_with_lists

  - file: tutorials/misc
    sections:
    - file: tutorials/misc_introduction_to_py5bot
    - file: tutorials/misc_jupyter_notebooks
    - file: tutorials/misc_rotating_cube
    - file: tutorials/misc_render_helper_tools
```

(I did some experimenting with Jupyter book but didn't test this exact config.)

When you use `glob:`, the files will be alphabetized. You'd need to add chapter numbers to the file names to control the ordering.  Listing out the pages with `file:` one at a time would also work.

The file `tutorials/intro_to_py5_and_python` would be an overview page where you could explain what the series of tutorials is about and maybe mention GSOC 2022, etc. Your tutorials would have names like `intro_to_py5_and_python_randomness` or `intro_to_py5_and_python_keyboard_and_mouse_events`. etc.

The other tutorials are a bit of a random bag of stuff that we can come up with a theme for. I'll take care of that and writing something for `tutorials/misc`. For now, dumping it in a subsection called misc is probably fine.

The Jupyter Book `sections:` are collapsible and are what are used in the Reference section for each of the py5 classes. I think it would keep this website neat and organized as the tutorials and how-tos grow.

If you make a PR for your tutorials we can get the TOC working well, I'll build this and put it on a dev site for us to look at and discuss.

**Reply by marcovicci** [reply]
I've gone ahead and used the first _toc.yml suggestion with `glob:` and (for now) I've only renamed a few of my files to match the scheme so we can see how it looks. (If the ordering ends up being more important to me I can manually list it out.) The `tutorials/intro_to_py5_and_python` page is sparse right now but I've filled it out with a few of my informational paragraphs from 'early' tutorials which suit it a bit better. Let me know if anything else needs to be addressed immediately in this one -- I'm a real greenhorn to contributing to projects on Github so I may have overlooked something in the PR! https://github.com/py5coding/py5book/pull/41

**Reply by hx2A** [reply]
**Important: I worked on this PR for much of today. There were a bunch of problems & headaches because of things I didn't know about jupyterbook and other stuff. I fixed the problems, but now have these ridiculously long messages on github that point out problems that get solved by the end. This might be a bit confusing to read, and it would be hard to correct it. I'm going to leave this so perhaps you can see my thought process and why certain changes were made. I recommend reading this but gloss over anything that is confusing, everything should be OK by the end.**

-------------

> I'm a real greenhorn to contributing to projects on Github so I may have overlooked something in the PR!

That's OK, I'm new at reviewing PRs of this size. We'll figure out how to do this together!

I was able to fetch the PR code and needed to make some changes (discussed below) to make the build work. It doesn't seem I can commit and push to your PR but I did create a new branch in the py5book repo called zellegsoc that contains your PR plus my additions.

https://github.com/py5coding/py5book/tree/zellegsoc

I think a good plan is for both of us to make PRs against the same branch that we both pull from. Is that OK with you? I can also make PRs against your fork of py5book if you prefer. Both will get the job done. Which is easier for you?

I uploaded this build to http://py5dev.ixora.io/. There are some problems that we will need to put our heads together and solve.

I explored these issues with the animation page, and you can see the results of my changes on that tutorial.

http://py5dev.ixora.io/tutorials/intro_to_py5_and_python_animation_and_motion.html

## Headache number 1

When jupyterbook runs the build, it will re-execute every notebook and store the results of each notebook cell. What appears in the build is its notebook executions, not yours. This will cause headaches for us because it will run one cell after another with no pauses. When you run the notebook, you will execute each cell with pauses, and for the cells that run a Sketch, you will let it run for a bit and then exit the Sketch. Unfortunately, jupyterbook does not know it needs to exit the Sketch. As a result, you will see this message:

```txt
Sketch is already running. To run a new sketch, exit the running sketch first.
```

in a lot of the notebooks (but not the animation one, because I fixed it there). The solve for this to explicitly call `exit_sketch()` after a small delay. That looks ugly when added to the documentation though. Happily, Myst Markdown has a solution for this that lets you add cells that either have their output hidden or cells that are completely removed.

https://jupyterbook.org/en/stable/reference/cheatsheet.html#myst-cheatsheet-code-cell-tags

In our case, you can add something like the following to the md files after `run_sketch()` gets called:

```md

```{code-cell} ipython3
:tags: ["remove-cell"]
import time  # this only needs to be done once per notebook

# give the Sketch some time to run and then exit
time.sleep(5)
exit_sketch()
time.sleep(0.25) # important to pause again because exiting is not instant
# (conclude this block with 3 backticks, but I can't type that in github)
```

The above cell will be executed but will not appear in the generated html page. This solves headache number 1, but there are problems if the Sketch happens to use print statements.

## Headache number 2

Print statements were a headache for py5 and Jupyter notebooks. Consider that a running py5 Sketch is using Java and is effectively multi-threaded. Q: if you run a Sketch with print statements in `draw()`, then move to a new cell and execute something simple in the new cell, where will the output of the print statements go? A: It will jump from the cell with the `run_sketch()` call and will now show up in the last executed cell. The output will follow you from one cell to the next.

Usually the solution for this is to use `println()`. This will "bind" to the cell that called `run_sketch()` and all output will only go there, regardless of what you do in the other cells. 

However, all of this works differently when jupyterbook is executing the notebook. If you use the remove-cell idea discussed above and are using print, the print output will go to the cell that is getting removed. Nothing will appear in the html output. If you use println, for some unknown reason the output gets mangled. I tried some other idea (that I don't like) that moves the sleep command to the cell with `run_sketch()` that you can see on this page:

http://py5dev.ixora.io/tutorials/intro_to_py5_and_python_animation_and_motion.html

With this, the output also gets mangled with the print output grouped together in groups of 7 to 14 lines. If I used println, it would be groups of 1. Why does it do this??? I have no idea. I've been working on this all afternoon and am currently feeling frustrated.

The print statements look great for http://py5dev.ixora.io/tutorials/intro_to_py5_and_python_randomness.html but perhaps would benefit from output scrolling. Use the tag "output_scroll" for that, as described in the myst cheatsheet. These print statements are from py5bot sketches, which for whatever reason jupyterbook can properly handle.

So what to do about print statements? I don't want to ask you to not use print statements in the documentation. I don't like that mangled output and adding `time.sleep()` again and again is weird and confusing. We are going to have to put our heads together and figure out what to do about this. Reading the jupyterbook documentation carefully and experimenting with different ideas will be necessary. We'll have to find some kind of path forward that works with Jupyterbook and provides the quality documentation that we are striving for.

**Reply by hx2A** [reply]
There are a few other changes that I made that were necessary to make this work, plus some other comments I should make.

* I added a tutorials/misc.md file to be a placeholder for the group of misc tutorials. Without that placeholder, they don't appear in the menu sidebar on the left.
* I added code to scripts/post_build.py so that your tutorials/images/ would get included in the build.
* You added a page to developer/ that talks about your involvement in GSoC. This is a good page and I am glad you wrote it. I think it might be better to put it at the end of your tutorials? Or we should create a new section? I don't know. Right now it is not in _toc.yml so it does not appear in the build.
* The first heading in each tutorial becomes the page title and should have one `#`, not two. I think this is an accessibility issue where pages are harder to navigate with a screen reader if the page skips a heading level? I believe Jupyterbook fixes this for you, but it issues a warning. I made this adjustment to the animation page but not the others.
* Only the pages that match tutorials/intro_to_py5_and_python_* are included in the build because only those are mentioned in _toc.yml. The other tutorials like working_with_text, etc, need to be renamed to be included. This seems like an intentional choice. Did you intend to rename them later?
* Your writing and documentation are so imaginative! This is fun to read. I'm so excited for what you are working on.

@marcovicci , there's a lot ahead of us but we will work together and solve the problems. Perseverance and teamwork are the driving force of open source projects. I know I wrote a lot here. If anything I wrote is confusing, I'd be happy to get a video call to explain stuff better. Let me know.

**Reply by hx2A** [reply]
I did some digging into the jupyterbook documentation and I found this:

https://jupyterbook.org/en/stable/content/execute.html#exclude-files-from-execution

Apparently Jupyter book doesn't have to execute each notebook itself; certain notebooks can be left alone. It will then create the page with the notebook as you saved it.

Here is the solution to Headache number 1 and number 2. If we put this in _config.yml:

```yml
execute:
  execute_notebooks: cache
  exclude_patterns:
    - 'intro_to_py5_and_python*'
```

It will skip the execution of the tutorials. Then you can have print statements with outputs that appear as you intended them to appear, without Jupyterbook executing everything again and making a mess. However, I tested this and it only works if you remove the myst markdown files. If the myst markdown files are there, it will use those for the build instead, and you'll have no outputs.

I just tested this with the animation notebook and it works great.

**Reply by hx2A** [reply]
OK! I now I have the other intro_to_py5_and_python* tutorials appearing correctly in the output, without jupyterbook re-executing them. Progress! I had to remove the myst markdown files. Headaches are solved!

There are maybe some cosmetic issues in the tutorials, like some of the notebook cells that have many rows of print statement output could benefit from the `:tags: ["output_scroll"]` tag, as described in https://jupyterbook.org/en/stable/reference/cheatsheet.html#myst-cheatsheet-code-cell-tags. You'll need to add this to the notebook cell metadata, since the myst files are removed.

Everything pushed to the zellegsoc branch of the py5book repo:

https://github.com/py5coding/py5book/tree/zellegsoc

**Reply by hx2A** [reply]
@marcovicci sorry if all of these messages are confusing!

Bottom line, here are the things you'll want to do next:

* Pull from the zellegsoc branch, but prioritize your copies of the tutorials over what is in that branch if something looks off
* Remove the Myst markdown files the notebooks are paired with
* Execute the tutorial notebooks and save them with the output in each cell as you'd like it to appear in the final documentation, if this has not been done already
* Consider adding the `"output_scroll"` tag to the notebook cells where you think it would be helpful
* Also, the `"raises-exception"` tag is useful for cells that are supposed to throw an error
* Rename tutorials to fit the intro_to_py5_and_python* pattern when you want them to be in the table of contents
* Make sure the first heading in each tutorial has one `#`, not two

That's enough writing for me today. I need a break from my computer!

**Comment by hx2A** [comment]
@marcovicci I hope you are having a good day today!

I just merged your PR into a new branch [zellegsoc](https://github.com/py5coding/py5book/tree/zellegsoc). Then I ran the build and deployed to the dev site:

http://py5dev.ixora.io/

Please pull from the zellegsoc branch and continue your work from there, and make the next PR against that. Moving forward, I'll approve your PRs as you make them so we both can have a current branch to work with.

A few things you'll notice on the dev site:

* In the tutorials section, there is a folding chapter for your intro tutorials. There's also a second folding chapter for the other tutorials, which I gave the generic name "misc". I'll improve that later.
* Not all of your tutorials are included in the intro tutorials chapter because not all of them match the glob pattern in the _toc.yml file `glob: tutorials/intro_to_py5_and_python_*`. If you rename those files to match the pattern, they will be included. However, the tutorials are probably not in the right order. Jupyter book will sort the globbed filenames lexicographically when doing the build. You can address this by inserting a strategically placed zero padded number (`01`, `02`, `03`, etc) to each filename or by listing out the filenames individually in _toc.yml.
* Your tutorials when rendered don't have page titles at the top. Compare the [iteration](file:///home/jim/Projects/ITP/pythonprocessing/py5book/_build/html/tutorials/intro_to_py5_and_python_iteration_with_loops.html) tutorial with the [introduction to py5bot](file:///home/jim/Projects/ITP/pythonprocessing/py5book/_build/html/tutorials/misc_introduction_to_py5bot.html) tutorial. To change that, you just need to change the first heading in each notebook to have one `#` instead of two (change `## iteration` to `# iteration`).

Let me know if you have any git issues and I will help you! Have a good day, and thank you for your hard work!

**Reply by hx2A** [reply]
The git commands I believe you'll need are:

```bash
git remote add upstream git@github.com:py5coding/py5book.git
git checkout -b zellegsoc upstream/zellegsoc
```

Alternatively, you can also create a new fork and uncheck the "fork only main branch" checkbox that github presents you. That will give you a new repo with the zellegsoc branch.

When you get a chance, try making a new PR to the zellegsoc branch with just the heading changes so we know we both have the git workflow stuff operating correctly.
